File: v1/src/main/java/com/api/v1/utils/CreditCardFinderUtil.java
package com.api.v1.utils;

import com.api.v1.domain.CreditCard;
import com.api.v1.domain.CreditCardRepository;
import com.api.v1.exceptions.CreditCardNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class CreditCardFinderUtil {

    @Autowired
    private CreditCardRepository repository;

    public Mono<CreditCard> find(String cardNumber) {
        return repository
                .findAll()
                .filter(e -> e.getCardNumber().equals(cardNumber))
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new CreditCardNotFoundException(cardNumber)));
    }

}

File: v2/src/main/kotlin/com/api/v2/creditcard/utils/CreditCardFinderUtil.kt
package com.api.v2.creditcard.utils

import com.api.v2.creditcard.domain.CreditCard
import com.api.v2.creditcard.domain.CreditCardRepository
import com.api.v2.creditcard.exceptions.CreditCardNotFoundException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class CreditCardFinderUtil {

    @Autowired
    lateinit var creditCardRepository: CreditCardRepository

    suspend fun find(cardNumber: String): CreditCard {
        return withContext(Dispatchers.IO) {
            val foundedCreditCard = creditCardRepository
                .findAll()
                .filter { e -> e.cardNumber == cardNumber }
                .firstOrNull();
            if (foundedCreditCard == null) throw CreditCardNotFoundException(cardNumber)
            foundedCreditCard
        }
    }

}